### PR TITLE
Docs for upcoming v2.5.0 release

### PR DESCRIPTION
**Pull request by francislavoie**
https://github.com/caddyserver/caddy/compare/v2.4.6...v2.5.0-beta.1, so far

TODO:

- [x] `tls` `get_certificate` (see #207)
- [x] `auto_https disable_certs`
- [x] `replace_status`
- [x] maybe update `remote_ip` matcher to mention zone support
- [x] `caddy trust` stuff, including API endpoints
- [x] `file_server` `pass_thru`
- [x] `trusted_proxies` stuff
- [x] dynamic upstreams
- [x] `dns_challenge_override_domain`
- [x] `reverse_proxy` `copy_response`

### CODE REVIEW

File: src/docs/markdown/caddyfile/directives.md [hunk 1]
```diff
@@ -49,6 +49,7 @@ Directive | Description
 **[import](/docs/caddyfile/directives/import)** | Include snippets or files
 **[log](/docs/caddyfile/directives/log)** | Enables access/request logging
 **[map](/docs/caddyfile/directives/map)** | Maps an input value to one or more outputs
+**[method](/docs/caddyfile/directives/method)** | Rewrite the HTTP method
```
Comment (mholt) [inline src/docs/markdown/caddyfile/directives.md]:
```suggestion
**[method](/docs/caddyfile/directives/method)** | Change the HTTP method internally
```

File: src/docs/markdown/caddyfile/directives/log.md [hunk 2]
```diff
@@ -160,9 +169,9 @@ format json
 
 #### single_field
 
-<span class="warning">⚠️ This format is deprecated, and will be removed in a future version.</span>
+<span class="warning">⚠️ This format is deprecated, and is removed in Caddy v2.5. To encode logs in common log format, please use the [`format-encoder`](https://github.com/caddyserver/format-encoder) plugin.</span>
```
Comment (mholt) [inline src/docs/markdown/caddyfile/directives/log.md]:
```suggestion
<span class="warning">⚠️ This format was deprecated, and has been removed in Caddy v2.5. To encode logs in common log format, please use the [`format-encoder`](https://github.com/caddyserver/format-encoder) plugin.</span>
```

File: src/docs/markdown/caddyfile/directives/log.md [hunk 3]
```diff
@@ -217,6 +225,59 @@ Masks IP addresses in the field using a CIDR mask, i.e. the number of bytes from
...
+
+##### regexp
+
+Marks a field to have a regular expression replacement applied at encoding time.
+
+```caddy-d
+<field> regexp <pattern> <replacement>
+```
+
+The regular expression language used is RE2, included in Go. See the [RE2 syntax reference](https://github.com/google/re2/wiki/Syntax) and the [Go regexp syntax overview](https://pkg.go.dev/regexp/syntax).
```
Comment (mholt) [inline src/docs/markdown/caddyfile/directives/log.md:267]: Good idea to link to these

File: src/docs/markdown/caddyfile/directives/log.md [hunk 4]
```diff
@@ -253,7 +314,7 @@ log {
 
 Use Common Log Format (CLF):
 
-<span class="warning">⚠️ The `single_field` format is deprecated and will be removed in a future version. To encode logs in common log format, please use the [`format-encoder`](https://github.com/caddyserver/format-encoder) plugin.</span>
+<span class="warning">⚠️ The `single_field` format is deprecated and removed in Caddy v2.5. To encode logs in common log format, please use the [`format-encoder`](https://github.com/caddyserver/format-encoder) plugin.</span>
```
Comment (mholt) [inline src/docs/markdown/caddyfile/directives/log.md]: Maybe we should just remove references to single_field and instead help people find what they need in the format-encoder directly.

File: src/docs/markdown/caddyfile/directives/method.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
+---
+title: method (Caddyfile directive)
+---
+
+# method
+
+Rewrites the request to change the HTTP method.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/directives/method.md]:
```suggestion
Changes the HTTP method on the request.
```

File: src/docs/markdown/caddyfile/directives/tls.md [hunk 6]
```diff
@@ -173,16 +173,19 @@ Note that ZeroSSL is a default issuer, so configuring it explicitly is usually u
 
 #### internal
 
-Obtains certificates from an internal certificate authority.
+Obtains certificates from an internal certificate authority.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/directives/tls.md]:
```suggestion
Obtains certificates from an internal certificate authority.
```

**Review by mholt — CHANGES_REQUESTED** [review]
These changes are _great_ so far. Thank you very much for your time and energy to work on these!

I'll be working on docs and the change log most of the rest of the day in the office today and maybe we can get the beta tagged!

File: src/docs/markdown/caddyfile/directives/log.md (hunk 3 above)
Comment (francislavoie) [reply src/docs/markdown/caddyfile/directives/log.md:267]: This is copied from all the other places we have regexp

File: src/docs/markdown/caddyfile/directives/log.md (hunk 4 above)
Comment (francislavoie) [reply src/docs/markdown/caddyfile/directives/log.md]: Yeah I wasn't sure whether we should leave a note in here or not

File: src/docs/markdown/caddyfile/directives/reverse_proxy.md [hunk 7]
```diff
@@ -48,8 +48,8 @@ Proxies requests to one or more backends with configurable transport, load balan
 ```caddy-d
 reverse_proxy [<matcher>] [<upstreams...>] {
     # backends
-    to <upstreams...>
-	...
+    to      <upstreams...>
+	dynamic <module> ...
```
Comment (francislavoie) [inline src/docs/markdown/caddyfile/directives/reverse_proxy.md]: mixed tabs and spaces in here

File: src/docs/markdown/caddyfile/directives/reverse_proxy.md [hunk 8]
```diff
@@ -128,6 +132,61 @@ If the address is not a URL (i.e. does not have a scheme), then placeholders can
 When proxying over HTTPS, you may need to override the `Host` header (which by default, retains the value from the original request) such that the `Host` header matches the TLS SNI value, which is used by servers for routing and certificate selection. See the [Headers](#headers) section below for more details.
 
 
+#### Dynamic upstreams
```
Comment (francislavoie) [inline src/docs/markdown/caddyfile/directives/reverse_proxy.md:148]: add this stuff to the TOC at the top

File: src/docs/markdown/caddyfile/directives/reverse_proxy.md [hunk 9]
```diff
@@ -128,6 +132,61 @@ If the address is not a URL (i.e. does not have a scheme), then placeholders can
...
+		dial_fallback_delay <duration>
+	}
+```
+
+- **&lt;name&gt;** - The full domain name of the record to look up (i.e. `_service._proto.name`).
+- **service** - The service component of the full name.
+- **proto** - The protocol component of the full name. Either `tcp` or `udp`.
+- **name** - The name component. Or, if `service` and `proto` are empty, the full domain name to query.
+- **refresh** - How often to refresh cached results. Default: `1m`
+- **resolvers** - List of resolvers to override system resolvers.
```
Comment (francislavoie) [inline src/docs/markdown/caddyfile/directives/reverse_proxy.md]: Should mention DNS on this line (in case people don't make the connection that resolving is for DNS)

File: src/docs/markdown/caddyfile/directives/reverse_proxy.md [hunk 10]
```diff
@@ -128,6 +132,61 @@ If the address is not a URL (i.e. does not have a scheme), then placeholders can
...
+- **dial_timeout** - Timeout for dialing the query.
+- **dial_fallback_delay** - Timeout for falling back from IPv6 to IPv6 via RFC 6555. Default: `300ms`
```
Comment (francislavoie) [inline src/docs/markdown/caddyfile/directives/reverse_proxy.md]: `IPv6 to IPv6` ?

File: src/docs/markdown/caddyfile/directives/reverse_proxy.md [hunk 11]
```diff
@@ -30,6 +30,10 @@ Proxies requests to one or more backends with configurable transport, load balan
 
 - [Syntax](#syntax)
 - [Upstreams](#upstreams)
+  - [Upstream addresses](#upstream-addresses)
+  - [Dynamic upstreams](#dynamic-upstreams)
+    - [SRV](#srv)
+	- [A/AAAA](#aaaaa)
```
Comment (francislavoie) [inline src/docs/markdown/caddyfile/directives/reverse_proxy.md]: there's a tab here instead of a space

**Comment by francislavoie** [comment]
I think that's it!

File: src/docs/markdown/api.md [hunk 12]
```diff
@@ -90,33 +96,33 @@ Export entire config and pretty-print it:
...
-									"handler": "file_server"
-								}
-							]
-						}
-					]
-				}
-			}
-		}
-	}
+  "apps": {
```
Comment (mholt) [inline src/docs/markdown/api.md]: Why the change to spaces? Tabs result in a small file and are user-configurable, and less error-prone.

File: src/docs/markdown/automatic-https.md [hunk 13]
```diff
@@ -179,6 +179,8 @@ The DNS challenge performs an authoritative DNS lookup for the candidate hostnam
 
 This challenge does not require any open ports, and the server requesting a certificate does not need to be externally accessible. However, the DNS challenge requires configuration. Caddy needs to know the credentials to access your domain's DNS provider so it can set (and clear) the special TXT records. If the DNS challenge is enabled, other challenges are disabled by default.
 
+Since ACME CAs follow DNS standard when looking up TXT records for challenge verification, you can use CNAME records to delegate answering the challenge to other DNS zones. This can be used to delegate the `_acme-challenge` subdomain to another zone. This is particularly useful if your DNS provider doesn't provide an API, or isn't supported by one of the DNS plugins for Caddy.
```
Comment (mholt) [inline src/docs/markdown/automatic-https.md]:
```suggestion
Since ACME CAs follow DNS standards when looking up TXT records for challenge verification, you can use CNAME records to delegate answering the challenge to other DNS zones. This can be used to delegate the `_acme-challenge` subdomain to another zone, which is particularly useful if your DNS provider doesn't provide an API or isn't supported by one of the DNS plugins for Caddy.
```

File: src/docs/markdown/caddyfile/directives/log.md [hunk 14]
```diff
@@ -13,6 +13,10 @@ $(function() {
 
 Enables and configures HTTP request logging (also known as access logs).
 
+<aside class="tip">
+	If you're looking to configure Caddy's runtime logs, you're looking for the <a href="/docs/caddyfile/options#log"><code>log</code> global option</a> instead.
```
Comment (mholt) [inline src/docs/markdown/caddyfile/directives/log.md:17]: Totally a nit, but does this make sense? I dunno what else to call them. Runtime is fine. I usually have called them Process logs. Or is there another word?

```suggestion
	If you're looking to configure Caddy's process logs, you're looking for the <a href="/docs/caddyfile/options#log"><code>log</code> global option</a> instead.
```

**Review by mholt — APPROVED** [review]
Looking good!

File: src/docs/markdown/api.md (hunk 12 above)
Comment (francislavoie) [reply src/docs/markdown/api.md]: Not user configurable in the docs though. Tab size is dictated by our CSS.

https://developer.mozilla.org/en-US/docs/Web/CSS/tab-size

I don't think the file size argument is relevant. Point is, it's easier to read deeply nested JSON with less width for each indentation level (I kinda want to make this change everywhere in the docs frankly)

File: src/docs/markdown/caddyfile/directives/log.md (hunk 14 above)
Comment (francislavoie) [reply src/docs/markdown/caddyfile/directives/log.md:17]: Both are essentially equivalent I guess. You could say "the Caddy runtime" or "the Caddy process", I think "runtime" might make more sense when talking about the lifecycle of Caddy, because a config != process' lifetime

File: src/docs/markdown/api.md (hunk 12 above)
Comment (mholt) [reply src/docs/markdown/api.md]: I think this change is very unrelated to v2.5 and should have separate discussion first. Maybe let's leave this out for consistency sake for now.

Comment (francislavoie) [reply src/docs/markdown/api.md]: ok
